eth/api_tracer_oe.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// OpenEthereumTrace A trace in the desired format (Parity/OpenEtherum) See: https://openethereum.github.io/wiki/JSONRPC-trace-module
+type OpenEthereumTrace struct {
+	Action              TraceRewardAction `json:"action"`
+	BlockHash           *common.Hash      `json:"blockHash"`
```

Comment:
Does this need to be a pointer? Since `BlockNumber` isn't, and neither type uses `omitempty`, it seems there is an assumption that the block context will always be available. Or?

Reply:
No it doesn't. Nice catch. It actually started wrong from wrong assumptions from OpenEthereum trace outputs and copying them from turbo-geth as starting structs.

Reply:
@meowsbits if we want to keep 100% compatibility with OpenEthereum tracer, then they return the below for rewards, as they use the same struct for both `trace_block` and `trace_transaction`.

```
    {
            "action": {
                "author": "0x877bd459c9b7d8576b44e59e09d076c25946f443",
                "rewardType": "block",
                "value": "0x4563918244f40000"
            },
            "blockHash": "0x8df024322173d225a09681d35edeaa528aca60743a11a70f854c158862bf5282",
            "blockNumber": 553416,
            "result": null,
            "subtraces": 0,
            "traceAddress": [],
            "transactionHash": null,
            "transactionPosition": null,
            "type": "reward"
        }
```

In our case where we use it only for block rewards, it's totally fine to remove the `TransactionHash` and `TransactionPosition`. What do you think? Shall we keep them?